### DISCUSSION TITLE
2.7.0 Release

### DISCUSSION THREAD

**Discussion by jtannas** [rswag/rswag#561] [Announcements]
## What's Changed
* drop support for EOL ruby + rails versions by @jtannas in https://github.com/rswag/rswag/pull/542
* update ChangeLog to replect contents of latest release by @jtannas in https://github.com/rswag/rswag/pull/549
* Make the docker user non-root for easier volume sharing by @jtannas in https://github.com/rswag/rswag/pull/550
* Update json-schema dependency version constraint by @mustardnoise in https://github.com/rswag/rswag/pull/517
* Fix Style/SingleArgumentDig rubocop issues in swagger_formatter by @rnestler in https://github.com/rswag/rswag/pull/486
* Update base path configuration for OpenAPI 3 by @romanblanco in https://github.com/rswag/rswag/pull/547
* make rspec-core dependency explicit by @jtannas in https://github.com/rswag/rswag/pull/554
* add deprecation notices for EOL Ruby & RSpec by @jtannas in https://github.com/rswag/rswag/pull/552
* Add Test Coverage Tooling by @jtannas in https://github.com/rswag/rswag/pull/551
* Override Content Security Policy for Swagger UI Compat by @eric-hemasystems in https://github.com/rswag/rswag/pull/263
* Update spec file path by @remy727 in https://github.com/rswag/rswag/pull/557
* Add ADDITIONAL_RSPEC_OPTS ENV var for rswag:specs:swaggerize by @asavageiv in https://github.com/rswag/rswag/pull/556
* chore: Update Url and QueryParam in cspell.json by @Jason3S in https://github.com/rswag/rswag/pull/560

## New Contributors
* @mustardnoise made their first contribution in https://github.com/rswag/rswag/pull/517
* @rnestler made their first contribution in https://github.com/rswag/rswag/pull/486
* @romanblanco made their first contribution in https://github.com/rswag/rswag/pull/547
* @eric-hemasystems made their first contribution in https://github.com/rswag/rswag/pull/263
* @remy727 made their first contribution in https://github.com/rswag/rswag/pull/557
* @Jason3S made their first contribution in https://github.com/rswag/rswag/pull/560

**Full Changelog**: https://github.com/rswag/rswag/compare/2.6.0...2.7.0

<hr /><em>This discussion was created from the release <a href='https://github.com/rswag/rswag/releases/tag/2.7.0'>2.7.0 Release Candidate</a>.</em>

**Comment by jtannas** [comment]
I've been wrapped for the last week with moving and still will be through this weekend.
Afaik the code on `master` is releasable but I haven't done any testing on it yet, so I've created a `2.7.0` git tag for the release candidate.

@sergioisidoro @romanblanco @RomainKoszyk 
If you'd be willing to help test it out then I'll publish once everything seems fine. Otherwise, I'll get to it soon™.

**Reply by sergioisidoro** [reply]
I can do some clicking around to make sure it's ok. I have master from git as a dependecy in a project and haven't noticed anything broken 👍

**Reply by RomainKoszyk** [reply]
Hi,

Thanks for the heads-up, I've pulled your RC and built the docker image. I can't get the CSP in the response headers, I assume it should have been added by doing so (maybe I'm missing something?):

```bash
curl -i http://127.0.0.1/index.html
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 07 Oct 2022 14:25:38 GMT
Content-Type: text/html
Content-Length: 3574
Last-Modified: Fri, 10 Dec 2021 14:34:35 GMT
Connection: keep-alive
Vary: Accept-Encoding
ETag: "61b3657b-df6"
X-Frame-Options: deny
Accept-Ranges: bytes

<!DOCTYPE html>
<!-- HTML for static distribution bundle build -->
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Swagger Editor</title>
  <style>
``` 

Otherwise, everything seems ok :)

**Reply by eric-hemasystems** [reply]
The CSP will only be returned if using the middleware. If you [copied the swagger ui to static files](https://github.com/rswag/rswag#serve-ui-assets-directly-from-your-web-server) and are having your web-server serve those then they would be missing.

That should be ok as the issue is when you configured the `content_security_policy.rb` file to be more restrictive than swagger needs is where the issue occurs. But static files served by your web server won't be following the `content_security_policy.rb` so no need to override it.

**Reply by tdeo** [reply]
Hello @jtannas, I'm working with @RomainKoszyk and trying to figure out installation of 2.7 from the Github release.

It looks like we'd need to install the .tar.gz archive directly, but I'm not sure if that's supported in Bundler / gemfile natively, otherwise would it be possible that you release this RC to rubygems directly?

Thanks

**Reply by jtannas** [reply]
When installing from github, there's an additional step needed to get the UI npm packages installed.
```bash
$ cd $(bundle show rswag-ui) && npm install
```

**Reply by tdeo** [reply]
Thanks for the command - by any chance would you happen to know or have any recommendations on how to automate this during the `bundle install` phase so that it works seemlessly in development environments and CI ?

**Reply by jtannas** [reply]
Not offhand :thinking: 
If you're not planning on running `master` as your rswag version long term then easiest way would be to just help test v2.7.0 so that we can release it.

**Reply by tdeo** [reply]
So I was just able to test it out, seems to work pretty well, I'm seeing a CSP violation with an image in the footer originating from there: https://github.com/swagger-api/swagger-ui/blob/32b3b97ae33635946d31711a3fcc94a6092b4e23/src/core/components/online-validator-badge.jsx#L58 

> Refused to load the image 'https://validator.swagger.io/validator?url=xxx' because it violates the following Content Security Policy directive: "img-src 'self' data:".

**Reply by eric-hemasystems** [reply]
My original PR had a allowance for that validator image but when I tested it against the latest code in master (which is using a newer version of swagger) that image seemed gone so I [removed it from the CSP](https://github.com/rswag/rswag/pull/263/commits/b663a996b5d8e070e522a8f95dcf05bcf5df37c8) when updating the PR.

Does rswag still render the older swagger interface for folks using the older swagger spec? If so we might need to add that allowance back.

**Reply by tdeo** [reply]
That's the image I'm not seeing on the very bottom right corner of the page:
<img width="1721" alt="image" src="https://user-images.githubusercontent.com/6230277/195188107-63d9fd91-231d-486a-b58d-0119ff6ef13a.png">

**Reply by eric-hemasystems** [reply]
Hmmm... that's a different URL. I didn't see that when I ran the test app but sounds like there might be some situation where that shows up so perhaps we need to add it?

**Reply by romanblanco** [reply]
I'm testing this in https://github.com/RedHatInsights/compliance-backend/pull/1455.

So far I have not noticed anything unexpected in the output :+1: 

Only @remy727's [change](https://github.com/rswag/rswag/pull/557) makes me wonder why was the path in `README.md` chaned.
In our project, the spec files are in `spec/integration`, so I'm thinking if we need to change something.

**Comment by gayatri-umesh** [comment]
Am having the issue with swagger UI rendering a blank page because of the CSP in our rails app being more restrictive than what swagger UI needs. I have updated rswag to 2.7.0 and still get the issue. 

Do I need to make changes in `application.rb` to use the middleware?

**Reply by eric-hemasystems** [reply]
You should already being mounting the middleware via the `routes.rb` file. Something like:

```ruby
mount Rswag::Ui::Engine => '/api-docs'
mount Rswag::Api::Engine => '/api-docs'
```

Unless you are serving the UI as static content but in that case the Rails CSP won't be an issue since the Rails CSP wouldn't be used.

**Reply by gayatri-umesh** [reply]
> You should already being mounting the middleware via the `routes.rb` file. Something like:
> 
> ```ruby
> mount Rswag::Ui::Engine => '/api-docs'
> mount Rswag::Api::Engine => '/api-docs'
> ```
> 
> Unless you are serving the UI as static content but in that case the Rails CSP won't be an issue since the Rails CSP wouldn't be used.

Yes am already doing that (have a custom one as below). 
```
mount Rswag::Ui::Engine => '/custom-api-docs'
mount Rswag::Api::Engine => '/custom-api-docs'
```
Still dont see the <meta> tag with the CSP applied when the index.html renders with the latest version `2.7.0`. Any other suggestions?
 ```
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Swagger UI</title>
  <link href="https://fonts.googleapis.com
```
One thing to point out is we use the [secure_headers](https://rubygems.org/gems/secure_headers/versions/2.1.0) gem.

**Reply by eric-hemasystems** [reply]
I don't know how secure_headers is implemented so that may be the reason. The way the CSP works in Rails is the CSP is only added if the response doesn't already have done. My patch had made it so rswag always provides one and therefore Rails doesn't need to add it's own.

secure_headers may not work that way. It may unconditionally add the CSP overwriting what rswag is doing. You would have to dig into the source code of secure_headers to know for sure.

**Reply by gayatri-umesh** [reply]
> I don't know how secure_headers is implemented so that may be the reason. The way the CSP works in Rails is the CSP is only added if the response doesn't already have done. My patch had made it so rswag always provides one and therefore Rails doesn't need to add it's own.
> 
> secure_headers may not work that way. It may unconditionally add the CSP overwriting what rswag is doing. You would have to dig into the source code of secure_headers to know for sure.

Yes my understanding is the same that `secure_headers` is overwriting and returning the application csp. With secure headers there is a [way](https://github.com/github/secure_headers/blob/main/docs/named_overrides_and_appends.md) to override, though, am not sure where to apply that in the rswag flow (usually applied in the controller layer). let me try if i could do it in `rswag_ui.rb` or `rswag_api.rb` to address this. I will reach out if I need any help. Thanks for helping with this.

**Reply by akaJSaunders** [reply]
@gayatri-umesh Im having the same issue that you are reporting here, have you made any progress in finding a solution?

**Comment by sthirugn** [comment]
hello folks, Is it possible to release 2.7.0 version soon? we have been waiting for a bugfix on this release, it will be very helpful if you can release this soon. Any help is much appreciated.

**Reply by asavageiv** [reply]
It was already released.

**Reply by sthirugn** [reply]
ah nice, I missed seeing it. Thanks @asavageiv for the info. Much appreciated.

**Comment by gayatri-umesh** [comment]
Hello Juliano,

I resolved the issue by moving the content security policy config from
`secure_headers` to rails CSP.

Hope this helps!

Thanks and Regards,
Gayatri

**Comment by epugh** [comment]
maybe just me, but on github, the most recent one that shows up on the homepage is 2.6????

**Comment by epugh** [comment]
ahh...  looks like 2.7 and 2.8 have a prerelease tag, which is why the homepage refers to 2.6!
